### PR TITLE
hostapd: convert wpa_psk_file items to sae_password parameters

### PR DESCRIPTION
**Pull request by rany2**
When SAE-only or SAE mixed mode is set and wpa_psk_file is enabled; SAE will not work without manually adding the sae_passphrase items to the hostapd config as well. This automatically handles the conversion of the wpa_psk_file file to the sae_passphrase analogues.

Only the pk parameter to sae_password doesn't have an analogue in wpa_psk_file however this shouldn't be needed for most users.

Limitations:

- WPA PSK file supports duplicate (mac,id) pairs but SAE does not. This is most problematic for a common use case where someone wants to have two wildcard entries that point to different VLANs (like a LAN and IoT).

### CODE REVIEW

**Review by raenye — APPROVED** [review]

**Comment by roshii** [comment]
I came upon the same issue but took a different approach to resolve it, actually leading to no know limitations.
See https://github.com/openwrt/openwrt/pull/14448

**Comment by rany2** [comment]
@roshii I suggest you wait until the sae_password_file parameter is added to hostapd that is available in OpenWRT. See https://w1.fi/cgit/hostap/commit/hostapd/hostapd.conf?id=e748e50c629f3e8d3c1c99843ef5439e568ed3ec

**Comment by rany2** [comment]
> actually leading to no know limitations

The limitation is fundamentally because SAE is different from WPA2. In SAE, if you set the same macaddr and keyId pair to some other config entry; it will override it. With WPA2, you can do something like:

```
00:00:00:00:00:00 password1
00:00:00:00:00:00 password2
```

and both password1 and password2 will be accepted. With SAE, this does not work and you will need to set a different keyid for password2; which is not well supported for any mainstream OS (Android, iOS, Windows, etc.)

**Comment by roshii** [comment]
> In SAE, if you set the same macaddr and keyId pair to some other config entry; it will override it
`mac` entry is optional in `sae_password`, so if mac address is `00:00:00:00:00:00` (catch-all) `mac` should not be set for SAE. so the limitation can easily be overcome with you approach as well.

> This is just so that the user doesn't have to set both WPA2 and SAE parameters.
It definitely make sense. I will therefore adapt my PR to make it a definitive solution for once sae_password_file parameter is added to hostapd that is available in OpenWRT.

**Comment by rany2** [comment]
> `mac` entry is optional in `sae_password`, so if mac address is `00:00:00:00:00:00` (catch-all) `mac` should not be set for SAE. so the limitation can easily be overcome with you approach as well.

That doesn't work. It behaves the same, in effect the default parameter when mac is not specified is `ff:ff:ff:ff:ff:ff`...

File: package/network/services/hostapd/files/hostapd.sh [hunk 1]
```diff
@@ -691,6 +691,60 @@ hostapd_set_bss_options() {
...
+							return 1
+						}
+
+						# Equivalent to 00:00:00:00:00:00 in WPA-PSK file is
+						# ff:ff:ff:ff:ff:ff in sae_password configuration.
+						[ "$macaddr" = "00:00:00:00:00:00" ] && macaddr="ff:ff:ff:ff:ff:ff"
+
+						# Generate the line for sae_password configuration and append it
+						# to the configuration file.
+						local sae_password_line="$psk|mac=$macaddr"
```
Comment (roshii) [inline package/network/services/hostapd/files/hostapd.sh:742]: With `sae_password` we can translate catch-all mac address "00:00:00:00:00:00" to no mac option, therefore allowing any client corresponding psk to authenticate.

**Review by roshii — CHANGES_REQUESTED** [review]

File: package/network/services/hostapd/files/hostapd.sh (hunk 1 above)
Comment (rany2) [reply package/network/services/hostapd/files/hostapd.sh:742]: See my comment here: https://github.com/openwrt/openwrt/pull/14343#issuecomment-1902719893

**Comment by roshii** [comment]
> That doesn't work. It behaves the same, in effect the default parameter when mac is not specified is `ff:ff:ff:ff:ff:ff`...

Works fine for me with the implementation from my PR. I can connect WPA3 clients and assign VLAN IDs without specifying any mac address

**Comment by rany2** [comment]
> Works fine for me with the implementation from my PR. I can connect WPA3 clients and assign VLAN IDs without specifying any mac address

Are you sure? The issue at hand is that you can't set the same SAE password for the same macaddr, keyid pair. So for example, if you had:

```
sae_password=password1|vlanid=1
sae_password=password2|vlanid=2
```

HostAPd will override `password1` and you will only be able to access VLAN 2 with password2. This is because, as previously stated the macaddr,keyid pairs are duplicates. HostAPd overrides the previous entry when that happens.

**Comment by roshii** [comment]
> Are you sure?

You are right, unfortunately... I could connect with the different password I had set only because the client I used kept the same mac address. Randomizing mac on the client causes authentication to fail.

**Review by roshii — APPROVED** [review]

**Comment by roshii** [comment]
> I could connect with the different password I had set only because the client I used kept the same mac address. Randomizing mac on the client causes authentication to fail.

Further testing demonstrate the following.

- When a client connects to the AP with the password set in the last sae_password entry it connects as intended and is assigned the correct VLAN ID.
- When the client, keeping the same mac address, connects to the AP using a random (!?) password, he is authenticated and assigned to the default network, as defined by wifi-iface config section.

@rany2 do you think the could be related to OpenWrt? I am suspecting hostap and would like to hear your opinion before raising the issue.

**Comment by rany2** [comment]
> When a client connects to the AP with the password set in the last sae_password entry it connects as intended and is assigned the correct VLAN ID.

This is expected per the hostapd documentation.

> When the client, keeping the same mac address, connects to the AP using a random (!?) password, he is authenticated and assigned to the default network, as defined by wifi-iface config section.

I really doubt this, how are you testing this?

**Comment by roshii** [comment]
> I really doubt this, how are you testing this?

I run an OpenWrt AP, including patches from my PR, with the following config.
I use an Android client, connecting using device mac address, first with guest password, the last sae_password entry, which receives an IP corresponding to guest VLAN. Then disconnecting, modifying password to a random one and reconnecting; the client then receives an IP corresponding to home network, the default AP network. Restarting hostapd via `wifi` command on AP prevent client to reconnect using a wrong password.

A logread on the AP show the following.
When connecting with valid password:
```
Mon Jan 22 10:52:09 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae IEEE 802.11: associated (aid 1)
Mon Jan 22 10:52:09 2024 daemon.notice hostapd: phy1-ap0: AP-STA-CONNECTED f2:0c:15:5f:43:ae auth_alg=sae
Mon Jan 22 10:52:09 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae RADIUS: starting accounting session EC4B4EA41D28B67C
Mon Jan 22 10:52:09 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae WPA: pairwise key handshake completed (RSN)
Mon Jan 22 10:52:09 2024 daemon.notice hostapd: phy1-ap0: EAPOL-4WAY-HS-COMPLETED f2:0c:15:5f:43:ae
```

Disconnecting:
```
Mon Jan 22 10:52:17 2024 daemon.notice hostapd: phy1-ap0: AP-STA-DISCONNECTED f2:0c:15:5f:43:ae
Mon Jan 22 10:52:17 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae IEEE 802.11: disassociated
Mon Jan 22 10:52:18 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae IEEE 802.11: deauthenticated due to inactivity (timer DEAUTH/REMOVE)
```

Reconnecting with a wrong password
```
Mon Jan 22 10:52:30 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae IEEE 802.11: authenticated
Mon Jan 22 10:52:30 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae IEEE 802.11: associated (aid 1)
Mon Jan 22 10:52:30 2024 daemon.notice hostapd: phy1-ap0: AP-STA-CONNECTED f2:0c:15:5f:43:ae auth_alg=open
Mon Jan 22 10:52:30 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae RADIUS: starting accounting session 045919528FDA63F5
Mon Jan 22 10:52:30 2024 daemon.info hostapd: phy1-ap0: STA f2:0c:15:5f:43:ae WPA: pairwise key handshake completed (RSN)
Mon Jan 22 10:52:30 2024 daemon.notice hostapd: phy1-ap0: EAPOL-4WAY-HS-COMPLETED f2:0c:15:5f:43:ae
```


Relevant OpenWrt config:
```
# cat /etc/config/wireless 

config wifi-device 'radio1'
	option type 'mac80211'
	option path 'platform/ahb/18100000.wmac'
	option channel '1'
	option band '2g'
	option country 'FR'
	option cell_density '0'
	option ht_coex '1'
	option htmode 'HT20'
	option txpower '14'

config wifi-iface 'buissonnets1'
	option device 'radio1'
	option mode 'ap'
	option ssid 'rocinante'
	option encryption 'sae-mixed'
	option ieee80211k '1'
	option wnm_sleep_mode '1'
	option wnm_sleep_mode_no_keys '1'
	option bss_transition '1'
	option proxy_arp '1'
	option multicast_to_unicast_all '1'
	option key 'secret127'
	option network 'home'

config wifi-vlan
	option name 'iot'
	option network 'iot'
	option vid '207'

config wifi-station
	option key 'secret207'
	option vid '207'

config wifi-vlan
	option name 'guest'
	option network 'guest'
	option vid '227'

config wifi-station
	option key 'secret227'
	option vid '227'
```

```
# cat /etc/config/network 

config interface 'loopback'
	option device 'lo'
	option proto 'static'
	option ipaddr '127.0.0.1'
	option netmask '255.0.0.0'

config globals 'globals'
	option ula_prefix 'fd33::/48'

config switch
	option name 'switch0'
	option reset '1'
	option enable_vlan '1'

config switch_vlan
	option device 'switch0'
	option vlan '127'
	option ports '0t 1t'
	option description 'home'
	option vid '127'

config switch_vlan
	option device 'switch0'
	option vlan '207'
	option ports '0t 1t'
	option description 'iot'
	option vid '207'

config switch_vlan
	option device 'switch0'
	option vlan '227'
	option ports '0t 1t'
	option description 'guest'
	option vid '227'

config device
	option type 'bridge'
	option name 'br-home'
	list ports 'eth0.127'

config device
	option type 'bridge'
	option name 'br-iot'
	list ports 'eth0.207'

config device
	option type 'bridge'
	option name 'br-guest'
	list ports 'eth0.227'

config interface 'home'
	option device 'br-home'
	option proto 'none'
	option delegate '0'

config interface 'iot'
	option device 'br-iot'
	option proto 'none'
	option delegate '0'

config interface 'guest'
	option device 'br-guest'
	option proto 'none'
	option delegate '0'
```

This results in the following hostapd.conf

```
# cat /tmp/run/hostapd-phy1.conf

driver=nl80211
logger_syslog=127
logger_syslog_level=2
logger_stdout=127
logger_stdout_level=2
country_code=FR
ieee80211d=1
hw_mode=g
supported_rates=60 90 120 180 240 360 480 540
basic_rates=60 120 240
beacon_int=100
channel=1
chanlist=1



#num_global_macaddr=1
ieee80211n=1
ht_coex=1
ht_capab=[SHORT-GI-20][SHORT-GI-40][TX-STBC][RX-STBC1][DSSS_CCK-40]

interface=phy1-ap0
ctrl_interface=/var/run/hostapd
ap_isolate=1
bss_load_update_period=60
chan_util_avg_period=600
disassoc_low_ack=1
skip_inactivity_poll=0
preamble=1
wmm_enabled=1
ignore_broadcast_ssid=0
uapsd_advertisement_enabled=1
utf8_ssid=1
multi_ap=0
sae_require_mfp=1
sae_pwe=2
wpa_passphrase=secret127
wpa_psk_file=/var/run/hostapd-phy1-ap0.psk
sae_password=secret207|vlanid=207
sae_password=secret227|vlanid=227
auth_algs=1
wpa=2
wpa_pairwise=CCMP
ssid=rocinante
bridge=br-home
wds_bridge=
snoop_iface=br-home
wnm_sleep_mode=1
wnm_sleep_mode_no_keys=1
bss_transition=1
rrm_neighbor_report=1
rrm_beacon_report=1
wpa_disable_eapol_key_retries=0
wpa_key_mgmt=WPA-PSK WPA-PSK-SHA256 SAE
okc=1
ieee80211w=1
group_mgmt_cipher=AES-128-CMAC
dynamic_vlan=0
vlan_naming=1
vlan_no_bridge=1
vlan_file=/var/run/hostapd-phy1-ap0.vlan
qos_map_set=0,0,2,16,1,1,255,255,18,22,24,38,40,40,44,46,48,56
multicast_to_unicast=1
proxy_arp=1
bssid=00:5f:67:46:f5:fe
#default_macaddr
```

**Comment by rany2** [comment]
I suggest you make a new issue about this, I can't reproduce it myself and don't see why it should be discussed in this PR.

**Comment by Neustradamus** [comment]
Linked to:
- https://github.com/openwrt/openwrt/pull/14343
- https://github.com/openwrt/openwrt/pull/14448
- https://github.com/openwrt/openwrt/pull/14515
- https://github.com/openwrt/openwrt/pull/14789
- https://github.com/openwrt/openwrt/pull/19964
- https://github.com/openwrt/openwrt/issues/19717
- https://github.com/openwrt/openwrt/pull/19718

**Pull request closed without merging**